src/main/java/ezvcard/io/chain/ChainingParser.java

```diff
@@ -141,31 +139,177 @@ public VCard first() throws IOException {
-		}
+		StreamReader reader = newReaderWithIndex();
 
 		try {
 			List<VCard> vcards = new ArrayList<>();
 			VCard vcard;
-			while ((vcard = reader.readNext()) != null) {
-				if (warnings != null) {
-					warnings.add(reader.getWarnings());
-				}
+			while ((vcard = readNextWithWarnings(reader)) != null) {
 				vcards.add(vcard);
 			}
 			return vcards;
 		} finally {
-			if (closeWhenDone()) {
-				reader.close();
-			}
-		}
+            closeIfNeeded(reader);
+        }
+	}
+
+	/**
+	 * Lazily reads vCards from the stream.
+	 * @return an {@link Iterator} over vCards being parsed
+	 */
+	@Override
+	public Iterator<VCard> iterator() {
+		return new ParsingIterator();
+    }
+
+	/**
+	 * Creates a {@link Stream} of vCards to be lazily read.
+	 * @return a {@link Stream} of vCards being parsed
+	 */
+	public Stream<VCard> stream() {
+		ParsingSpliterator spliterator = new ParsingSpliterator();
+		return StreamSupport.stream(spliterator, false).onClose(spliterator::close);
 	}
 
 	abstract StreamReader constructReader() throws IOException;
 
 	private boolean closeWhenDone() {
 		return in == null && reader == null;
 	}
+
+	private StreamReader newReaderWithIndex() throws IOException {
+		StreamReader reader = constructReader();
+		if (index != null) {
+			reader.setScribeIndex(index);
+		}
+		return reader;
+	}
+
+	private VCard readNextWithWarnings(StreamReader reader) throws IOException {
+		VCard vCard = reader.readNext();
+		if (warnings != null) {
+			warnings.add(reader.getWarnings());
+		}
+		return vCard;
+	}
+
+	private void closeIfNeeded(StreamReader reader) throws IOException {
+		if (closeWhenDone()) {
+			reader.close();
+		}
+	}
+
+	private final class ParsingIterator extends Parserator implements Iterator<VCard> {
+		private VCard next;
+
+		@Override
+		public boolean hasNext() {
+			return !done && (next != null || (next = advance()) != null);
+		}
+
+		@Override
+		public VCard next() {
+			if (done) {
+				throw new NoSuchElementException();
+			}
+
+			VCard next = this.next;
+			if (next != null) {
+				this.next = null;
+				return next;
+			}
+
+			if ((next = advance()) != null) {
+				return next;
+			}
+
+			throw new NoSuchElementException();
+		}
+	}
+
+	private final class ParsingSpliterator extends Parserator implements Spliterator<VCard> {
+		@Override
+		public boolean tryAdvance(Consumer<? super VCard> action) {
+			if (done) {
+				return false;
+			}
+
+			VCard vCard = advance();
+			if (vCard != null) {
+				action.accept(vCard);
+				return true;
+			}
+
+			return false;
+		}
+
+		@Override
+		public Spliterator<VCard> trySplit() {
+			return null;
+		}
+
+		@Override
+		public long estimateSize() {
+			return Long.MAX_VALUE;
+		}
+
+		@Override
+		public int characteristics() {
+			return Spliterator.ORDERED | Spliterator.NONNULL | Spliterator.IMMUTABLE;
+		}
+	}
+
+	private class Parserator {
+		private StreamReader reader;
+		boolean done = false;
+
+		Parserator() {
```

Comment:
Any reason for this constructor? Can it be removed?

Reply:
Looks like it's useless, just a habit to specify visibility explicitly.
Done.